### PR TITLE
Add sign-ext lowering pass when target browser doesn't support it

### PR DESCRIPTION
**Pull request by sbc100**
llvm and binaryen now default to enabling sign-ext and mutable-globals   
by default. See https://reviews.llvm.org/D125728.                        
                                                                         
In order to allow users to continue to target older browsers we will        
now lower sign-ext operations in a binaryen pass if the target browsers  
don't support it.

### CODE REVIEW

**Comment by sbc100** [comment]
This is really a WIP to show what we will likely have to do in preparation for the llvm-side change.

**Comment by dschuff** [comment]
```

**Comment by dschuff** [comment]
```
As part of this change we are bumping the default minimum browser
versions for Safari and Firefox to the minimum version needed to
supported these 4 features which are about to be enabled by default:
```
should we do a no-op change first, which doesn't bump the versions?

Another question. I assume that if a user passes `-mnontrapping-fptoint` or `-pthreads` or `-fwasm-exceptions` or `-mno-signext` that it will have the usual effect, independently of this change?
(actually threads and EH are special since I guess they'll stay off by default anyway....) But in other words, if a user uses a clang flag it can override what emscripten thinks the default feature set is. this is fine if the feature is only enabled by clang, but if a feature also needs support in emscripten it gets more complicated...

**Comment by sbc100** [comment]
> Another question. I assume that if a user passes `-mnontrapping-fptoint` or `-pthreads` or `-fwasm-exceptions` or `-mno-signext` that it will have the usual effect, independently of this change?
> (actually threads and EH are special since I guess they'll stay off by default anyway....) But in other words, if a user uses a clang flag it can override what emscripten thinks the default feature set is. this is fine if the feature is only enabled by clang, but if a feature also needs support in emscripten it gets more complicated...

Yes that works both before and after this change explicit user flags come later and therefore take precedence

**Comment by tlively** [comment]
New feature matrix code looks good to me 👍

**Comment by sbc100** [comment]
I split out the actual bumping the default versions into #17690

**Comment by sbc100** [comment]
Added a test. PTAL

File: tools/feature_matrix.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    'safari': 141000,
+  },
+  Feature.BULK_MEMORY: {
+    'chrome': 75,
+    'firefox': 79,
+    'safari': 150000,
+  },
+  Feature.MUTABLE_GLOBALS: {
+    'chrome': 74,
+    'firefox': -1, # Firefox has always supported mutable globals apparently
```
Comment (dschuff) [inline tools/feature_matrix.py:41]: I found https://bugzilla.mozilla.org/show_bug.cgi?id=1412238, it looks like version 61

**Review by dschuff — COMMENTED** [review]
I guess we could also ask the opposite question of "what if the user only passes the feature flags at link time and doesn't at compile time": I guess in that case things would also work as they do now... wasm features would be unaffected, but JS/web features would be adjusted accordingly.

File: tools/settings.py [hunk 2]
```diff
@@ -69,6 +69,13 @@
     'WASM_OBJECT_FILES',
     'WASM_WORKERS',
 
+    # These are needed at compile time to calulate feature flags
```
Comment (kripken) [inline tools/settings.py:72]:
```suggestion
    # These are needed at compile time to calculate feature flags
```

File: test/test_other.py [hunk 3]
```diff
@@ -10612,14 +10612,38 @@ def fail(args, details):
...
+      return [x[1] for x in features if x[0] == '+']
+
+    # By default we expect to see some features disabled since our min browser versions
+    # don't allow them.
+    self.run_process([EMCC, '-mcpu=bleeding-edge', '-c', test_file('hello_world.c')])
+    enabled = get_enabled_features('hello_world.o')
+    self.assertNotContained('nontrapping-fptoint', enabled)
+    self.assertNotContained('mutable-globals', enabled)
+
+    # If we not disable all support for Edge/Safari/Firefox those feature should now be
```
Comment (kripken) [inline test/test_other.py:10637]:
```suggestion
    # If we disable all support for Edge/Safari/Firefox those feature should now be
```

File: test/test_other.py [hunk 4]
```diff
@@ -10612,14 +10612,38 @@ def fail(args, details):
...
+    # enabled by the -mcpu=bleeding-edge.
```
Comment (kripken) [inline test/test_other.py:10638]:
```suggestion
    # enabled by -mcpu=bleeding-edge.
```

File: test/test_other.py [hunk 5]
```diff
@@ -10612,14 +10612,38 @@ def fail(args, details):
...
+    self.run_process([EMCC, '-mcpu=bleeding-edge', '-c', test_file('hello_world.c'),
+                      '-sMIN_EDGE_VERSION=-1',
+                      '-sMIN_SAFARI_VERSION=-1',
+                      '-sMIN_FIREFOX_VERSION=-1'])
```
Comment (kripken) [inline test/test_other.py:10642]: Why is Chrome different here?

Comment (sbc100) [reply test/test_other.py:10643]: Because its the default values of the other browsers that prevent those features being enabled.  The default chrome version is good enough apparently.  I updated the comment.

File: test/test_other.py [hunk 6]
```diff
@@ -10612,14 +10612,39 @@ def fail(args, details):
...
+      return [x[1] for x in features if x[0] == '+']
+
+    # By default we expect to see some features disabled since our min browser versions
+    # don't allow them.
+    self.run_process([EMCC, '-mcpu=bleeding-edge', '-c', test_file('hello_world.c')])
+    enabled = get_enabled_features('hello_world.o')
+    self.assertNotContained('nontrapping-fptoint', enabled)
+    self.assertNotContained('mutable-globals', enabled)
+
+    # If we not disable all support for Edge/Safari/Firefox (the browser versions whose
```
Comment (kripken) [inline test/test_other.py:10637]:
```suggestion
    # If we disable all support for Edge/Safari/Firefox (the browser versions whose
```

File: test/test_other.py [hunk 7]
```diff
@@ -10612,14 +10612,39 @@ def fail(args, details):
...
+    # default values preclude bleeding edge features) those feature should now be enabled
+    # by when using -mcpu=bleeding-edge.
+    self.run_process([EMCC, '-mcpu=bleeding-edge', '-c', test_file('hello_world.c'),
+                      '-sMIN_EDGE_VERSION=-1',
+                      '-sMIN_SAFARI_VERSION=-1',
+                      '-sMIN_FIREFOX_VERSION=-1'])
+    enabled = get_enabled_features('hello_world.o')
+    self.assertContained('nontrapping-fptoint', enabled)
+    self.assertContained('mutable-globals', enabled)
+
```
Comment (kripken) [inline test/test_other.py:10647]: is there a feature we could test here that *is* enabled by default and is not in the MVP?

File: tools/feature_matrix.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    return False
+  return True
+
+
+def get_feature_flags():
+  flags = []
+
+  # Enabling threads, or PIC implicitly opts into many of these features.
+  # TODO(sbc): We should probably have USE_PTHREADS/RELOCATABLE
+  # implicitly bump the min browser versions instead.
```
Comment (kripken) [inline tools/feature_matrix.py:66]: I think this TODO would be a good idea. Here is where we do it already for bigint:

https://github.com/emscripten-core/emscripten/blob/0fe8aa8e91a77aea12068e31ff4c5d3ee44a497c/emcc.py#L401-L413

**Review by kripken — COMMENTED** [review]
Please mention this in the changelog. Would be good to mention both directions - both that we disable as needed, and enable based on pthreads and bulk memory now.

File: tools/feature_matrix.py (hunk 8 above)
Comment (dschuff) [reply tools/feature_matrix.py:66]: This is somewhat surprising behavior (not the linking of pthreads with dependent features, but the upgrading of all features because one of them requires a particular browser version). It sort of makes sense if the user is targeting a known browser, but there could be use cases where you'd want to let the user override emscripten's choices. Most of them are variations of "the user knows more about the feature matrix than emscripten does." for example:  using a browser (or other JS runtime) that is unknown to emscripten, using an environment that has features removed for security, or even added (something like electron?). All of these might have different feature sets so the user would probably either pick a browser baseline and add features to it, or just manually select features. (I guess it also conflicts with the behavior that adding clang flags like `-mmultivalue` does not affect emscripten's choice of target versions)

File: tools/feature_matrix.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    logger.debug('adding -mnosign-ext due to target browser selection')
+    flags.append('-mno-sign-ext')
+
+  if not caniuse(Feature.BULK_MEMORY):
+    logger.debug('adding -mnobulk-memory due to target browser selection')
+    flags.append('-mno-bulk-memory')
+
+  if not caniuse(Feature.MUTABLE_GLOBALS):
+    logger.debug('adding -mnomutable-globals due to target browser selection')
+    flags.append('-mno-mutable-globals')
```
Comment (juj) [inline tools/feature_matrix.py:84]: It might make sense to promote these debug messages to warnings? And if there is an easy way to make them more descriptive, something along the lines of 
```
Requested MIN_SAFARI_VERSION=120000, where Wasm feature "nontrapping-fptoint" is not supported. Building with -mno-nontrapping-fptoint to disable targeting this feature. (-Wunsupported-wasm-feature)
```
This way user would be able to promote the scenario from a warning to an error with a `-Werror=unsupported-wasm-feature` or mute it with a `-Wno-unsupported-wasm-feature`, so be able to get explicit behavior checks out of the compiler?

This way the compiler would not risk accidentally disabling a feature possibly without the user knowing? (user might not realize to run with EMCC_DEBUG enabled to check the logs)

Comment (sbc100) [reply tools/feature_matrix.py:84]: Good idea!

File: emcc.py [hunk 10]
```diff
@@ -613,6 +614,13 @@ def get_binaryen_passes():
     passes += ['--safe-heap']
   if settings.MEMORY64 == 2:
     passes += ['--memory64-lowering']
+  # sign-ext is enabled by default by llvm.  If the target browser settings don't support
+  # this we lower it away here using a binaryen pass.
+  if not feature_matrix.caniuse(feature_matrix.Feature.SIGN_EXT):
+    logger.debug('lowering sign-ext proposal due to incompatiable target browser engines')
+    passes += ['--signext-lowering']
+  if settings.MEMORY64 == 2:
+    passes += ['--memory64-lowering']
```
Comment (kripken) [inline emcc.py:623]: These two lines appear above as well - bad copy-paste?

File: test/test_other.py [hunk 11]
```diff
@@ -12724,3 +12724,22 @@ def test_reproduce(self):
     response = response.replace('\\', '/')
     response = response.replace(root, '<root>')
     self.assertTextDataIdentical(expected, response)
+
+  def test_signext_lowering(self):
+    cmd = [EMCC, test_file('other/test_signext_lowering.c'), '-sWASM_BIGINT', '-sMIN_EDGE_VERSION=0x7fffffff']
+
+    # We expect ERROR_ON_WASM_CHANGES_AFTER_LINK to fail due lowering of sign-ext being
+    # required for safary 12.
```
Comment (kripken) [inline test/test_other.py:12732]:
```suggestion
    # required for safari 12.
```

File: tools/feature_matrix.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+
+def caniuse(feature):
+  min_versions = min_browser_versions[feature]
+  if settings.MIN_CHROME_VERSION < min_versions['chrome']:
+    return False
+  if settings.MIN_FIREFOX_VERSION < min_versions['firefox']:
+    return False
+  if settings.MIN_SAFARI_VERSION < min_versions['safari']:
+    return False
```
Comment (kripken) [inline tools/feature_matrix.py:56]: The test checks safari, but maybe worth testing the other browser flags too, that is, to get test coverage for the lines with chrome, firefox, ie, and edge here?

**Review by kripken — COMMENTED** [review]
lgtm % comments

File: emcc.py [hunk 13]
```diff
@@ -613,6 +614,11 @@ def get_binaryen_passes():
     passes += ['--safe-heap']
   if settings.MEMORY64 == 2:
     passes += ['--memory64-lowering']
+  # sign-ext is enabled by default by llvm.  If the target browser settings don't support
+  # this we lower it away here using a binaryen pass.
+  if not feature_matrix.caniuse(feature_matrix.Feature.SIGN_EXT):
+    logger.debug('lowering sign-ext proposal due to incompatiable target browser engines')
```
Comment (juj) [inline emcc.py:620]: `incompatiable` -> `incompatible`

File: test/test_other.py [hunk 14]
```diff
@@ -12724,3 +12724,32 @@ def test_reproduce(self):
...
+    output = self.expect_fail(cmd + ['-sERROR_ON_WASM_CHANGES_AFTER_LINK', '-sMIN_SAFARI_VERSION=120000'])
+    self.assertContained('error: changes to the wasm are required after link, but disallowed by ERROR_ON_WASM_CHANGES_AFTER_LINK', output)
+    self.assertContained('--signext-lowering', output)
+
+    # Same for firefox
+    output = self.expect_fail(cmd + ['-sERROR_ON_WASM_CHANGES_AFTER_LINK', '-sMIN_FIREFOX_VERSION=61'])
+    self.assertContained('error: changes to the wasm are required after link, but disallowed by ERROR_ON_WASM_CHANGES_AFTER_LINK', output)
+    self.assertContained('--signext-lowering', output)
+
+    # Same for firefox
```
Comment (juj) [inline test/test_other.py:12742]: # Same for chrome

**Comment by sbc100** [comment]
> If I understand correctly:
> 
> 1. LLVM still supports compiling code with both sign-ext enabled vs disabled, and mutable-globals enabled vs disabled. So they did not e.g. drop support altogether for compiling against sign-ext disabled for example?

Correct.

> 2. The way this PR is laid out is that if user passes MIN_xxx_VERSION smaller than what is required for sign-ext, an extra Binaryen pass is run to lower the sign-ext code, and a log diagnostic message is printed to the `debug` channel that this happened. This Binaryen lowering pass occurs even if user had passed `-mno-sign-ext` to compile+link stages in the first place?

Current.  The lowering pass happens when you target older browsers, regardless of whether `-mno-sign-ext` was used to try to avoid the instruction.

One compelling reason for doing it this way is it ensuring that all input files are built with `-mno-sign-ext` isn't just about user code, it would require all system libraries to be rebuilt with this flag (i.e. adding a new system library variant)

> 
> Do the Wasm object files record which features had been enabled when they were being built? That would allow quickly detecting if a lowering pass is needed at all?

Yes, wasm object files do contain that information, and actually I think we could rely on the linker outputting the features needed in the final binary.

However, I'm not sure its worth doing this, since the running of this pass is only done if a non-default set of browsers are targeted.    If such older browser are being targeted then is almost certain that `-sWASM_BIGINT` is also not being used, so binaryen is already required anyway, and adding this pass is basically free.

**Comment by sbc100** [comment]
> and a log diagnostic message is printed to the `debug` channel that this happened.

Do you think reporting this information is important?  I don't think we report it when we perform i64 lowering, which happens in the default case (i.e. when `-sWASM_BIGINT` is not passed).

File: emcc.py (hunk 13 above)
Comment (dschuff) [reply emcc.py:620]: Also if I were nitpicking i'd call this a "feature" rather than a "proposal" since it's fully merged into the spec :D

File: emcc.py [hunk 15]
```diff
@@ -476,7 +477,6 @@ def default_min_browser_version(browser, version):
 
   if settings.WASM_BIGINT:
     default_min_browser_version('Safari', 150000)
-    default_min_browser_version('Edge', 79)
```
Comment (kripken) [inline emcc.py:479]: unrelated change?

**Review by kripken — APPROVED** [review]

File: emcc.py (hunk 15 above)
Comment (sbc100) [reply emcc.py:479]: Actually its needed, since setting the edge version to anything by -1 currently means no features are supported.  i.e. the edge version currently means only EdgeHTML and not modern Edge.

Comment (sbc100) [reply emcc.py:479]: The proposal is that for now the edge version only refers to the old edge browser, and if the features of modern edge ever diverge from chromium we can start to support `MIN_EDGE_VERSION` being distinct from `MIN_CHROME_VERSION`

Not mentioning edge at all there just means that default of "-1" (not supported at all) will prevail.

Right now I'm not considering edge at all in the feature_matix.py because we currently assume that folks can use the MIN_CHROMIUM_VERSION to target modern versions of edge.    Instead of removing this line I could instead extend feature_matrix.py to include support for MIN_EDGE_VERSION (but it would just be identical what is already there for chromium).

Comment (sbc100) [reply emcc.py]: OK, I reworked `feature_matrix.py` such that this change is no longer needed.

Comment (kripken) [reply emcc.py:479]: Oh, then the Chrome version refers to modern Edge too? Sgtm, but I wasn't aware their features had even been in sync so far. Is that the case in practice?

**Review by kripken — APPROVED** [review]

File: emcc.py (hunk 15 above)
Comment (sbc100) [reply emcc.py:479]: If you looks at the versions in https://en.wikipedia.org/wiki/Microsoft_Edge (expand "New Edge release history") you will see that edge versions correspond 1-to-1 with blink versions.. to from a JS/wasm perspective one would expect them to always match.

**Comment by juj** [comment]
> it would require all system libraries to be rebuilt with this flag

Good point, forgot about that. Agree, we wouldn't want to build different variants of the system libs for each combination.

**Comment by juj** [comment]
> If such older browser are being targeted then is almost certain that -sWASM_BIGINT is also not being used, so binaryen is already required anyway, and adding this pass is basically free.

Ah right, there's already other steps that are needed. Yeah sgtm like this.

**Comment by Alexufo** [comment]
The same. Is there any news?
